lib/chef/resource/aws_route53_hosted_zone.rb

```diff
@@ -0,0 +1,220 @@
+require 'chef/provisioning/aws_driver/aws_resource'
+require 'chef/resource/aws_route53_record_set'
+require 'securerandom'
+
+# the AWS API doesn't have these objects linked, so give it some help.
+class Aws::Route53::Types::HostedZone
+  attr_accessor :resource_record_sets
+end
+
+class Chef::Resource::AwsRoute53HostedZone < Chef::Provisioning::AWSDriver::AWSResourceWithEntry
+
+  aws_sdk_type ::Aws::Route53::Types::HostedZone, load_provider: false
+
+  resource_name :aws_route53_hosted_zone
+
+  # name of the domain. AWS will tack on a trailing dot, so we're going to prohibit it here for consistency:
+  # the name is our data bag key, and if a user has "foo.com" in one resource and "foo.com." in another, Route
+  # 53 will happily accept two different domains it calls "foo.com.".
+  attribute :name, kind_of: String, callbacks: { "domain name cannot end with a dot" => lambda { |n| n !~ /\.$/ } }
+
+  # The comment included in the CreateHostedZoneRequest element. String <= 256 characters.
+  attribute :comment, kind_of: String
+
+  attribute :aws_route53_zone_id, kind_of: String, aws_id_attribute: true
+
+  def record_sets(&block)
+    if block_given?
+      @record_sets_block = block
+    else
+      @record_sets_block
+    end
+  end
+
+  def aws_object
+    driver, id = get_driver_and_id
+    result = driver.route53_client.get_hosted_zone(id: id).hosted_zone if id rescue nil
+    if result
+      result.resource_record_sets = get_record_sets_from_aws(result.id).resource_record_sets
+      result
+    else
+      nil
+    end
+  end
+
+  # since this is used exactly once, it could plausibly be inlined in #aws_object.
+  def get_record_sets_from_aws(hosted_zone_id, opts={})
+    params = { hosted_zone_id: hosted_zone_id }.merge(opts)
+    driver.route53_client.list_resource_record_sets(params)
+  end
+end
+
+class Chef::Provider::AwsRoute53HostedZone < Chef::Provisioning::AWSDriver::AWSProvider
```

Comment:
That happened when initially pairing with Mr. Keiser way back when, this is definitely part of his desired erasure of the resource/provider split. I don't know that I've seen a roadmap for that larger direction, though.